### PR TITLE
Feature/api documentation improvement

### PR DESCRIPTION
**Pull request by rfeistenauer**
## Description

These changes improve the API documentation by clarifying the responsibilities of classes and the class structure.

### Things to add, update or check before this PR can be merged.

* [x] Public api function documentation
* [x] [Tutorials](https://wiki.ros.org/pilz_robots/Tutorials/)
* [x] Overview on [ROS wiki](https://wiki.ros.org/pilz_robots)
* [x] Package Readme ([example pilz_robots](https://github.com/PilzDE/pilz_robots/blob/melodic-devel/README.md))
* [x] Good commit messages ([some tips](https://dev.to/jacobherrington/how-to-write-useful-commit-messages-my-commit-message-template-20n9
))
* [x] CHANGELOG.rst updated
* [x] Copyright headers
* [x] Examples

### Review Checklist
* [x] Soft- and hardware architecture (diagrams and description)
* [x] Test review (test plan and individual test cases)
* [x] Documentation review
* [x] Code (coding rules, style guide)

### Release planning (please answer)
* [x] When is the new feature released?
Needs an release to update the online documentation but is not urgent?

* [x] Which dependent packages do have to be released simultaneously?
None

### CODE REVIEW

File: include/psen_scan_v2/monitoring_frame_deserialization.h [hunk 1]
```diff
@@ -71,8 +88,25 @@ class FixedFields
...
+ * of a monitoring frame.
+ *
+ * @see monitoring_frame
+ */
 namespace additional_field
 {
+/**
+ * @brief Definition for the type and length of an additional field in a monitoring frame.
+ *
+ * The exact content of an monitoring frame differs depending on the configuration.
```
Comment (cdoehn) [inline include/psen_scan_v2/monitoring_frame_deserialization.h:102]:
```suggestion
 * The exact content of a monitoring frame differs depending on the configuration.
```

File: include/psen_scan_v2/monitoring_frame_deserialization.h [hunk 2]
```diff
@@ -71,8 +88,25 @@ class FixedFields
   FromTheta from_theta_;
   Resolution resolution_;
 };
+/**
+ * @brief Namespace containing all informations/types needed to describe the content of the additional information field
```
Comment (cdoehn) [inline include/psen_scan_v2/monitoring_frame_deserialization.h:92]:
```suggestion
 * @brief Namespace containing all information/types needed to describe the content of the additional information field
```

File: include/psen_scan_v2/monitoring_frame_deserialization.h [hunk 3]
```diff
@@ -24,6 +24,20 @@
 
 namespace psen_scan_v2
 {
+/**
+ * @brief Namespace defining all informations/types needed to describe a monitoring frame.
```
Comment (cdoehn) [inline include/psen_scan_v2/monitoring_frame_deserialization.h:28]:
```suggestion
 * @brief Namespace defining all information/types needed to describe a monitoring frame.
```

File: include/psen_scan_v2/monitoring_frame_deserialization.h [hunk 4]
```diff
@@ -24,6 +24,20 @@
...
+ *
+ * A complete scan round can be divided in several parts sent independently over the UDP connection.
+ * Those parts are called monitoring frames.
+ *
+ * Every single frame **has to** contain some general information about the scan in the fixed fields
+ * and **can** contain additional data like distances or intensitis in additional fields.
```
Comment (cdoehn) [inline include/psen_scan_v2/monitoring_frame_deserialization.h:34]:
```suggestion
 * and **can** contain additional data like distances or intensities in additional fields.
```

File: include/psen_scan_v2/ros_scanner_node.h [hunk 5]
```diff
@@ -34,7 +34,7 @@ namespace psen_scan_v2
 using namespace std::chrono_literals;
 
 /**
- * @brief ROS Node for fetching and publishing laserscan data from the scanner.
+ * @brief ROS Node that continuesly publishes scan data of a single PSENscan module.
```
Comment (cdoehn) [inline include/psen_scan_v2/ros_scanner_node.h:37]:
```suggestion
 * @brief ROS Node that continuesly publishes scan data of a single PSENscan laser scanner.
```

File: include/psen_scan_v2/scanner_interface.h [hunk 6]
```diff
@@ -26,7 +26,12 @@
 namespace psen_scan_v2
 {
 /**
- * @brief API to control and to fetch measurements from the scanner.
+ * @brief This is the API definition for external interaction with the scanner driver.
+ *
+ * This Interface allows to:
```
Comment (cdoehn) [inline include/psen_scan_v2/scanner_interface.h:31]:
```suggestion
 * This interface allows to:
```

**Review by cdoehn — CHANGES_REQUESTED** [review]
Additionaly, I think the toLaserScan function should be documented, because it owns an important functionality

File: doc/calls_and_callbacks.md [hunk 7]
```diff
@@ -1 +0,0 @@
-![psen_scan_v2_callbacks](api_sequence_diag.png)
```
Comment (ct2034) [inline doc/calls_and_callbacks.md:1]: what is this? where does it come from, where does it go?

File: doc/diag_comp_overall_architecture.uxf [hunk 8]
```diff
@@ -42,347 +67,405 @@ UDP package</panel_attributes>
...
     </coordinates>
     <panel_attributes>symbol=component
 style=wordwrap
+layer=-10
 &lt;&lt;User API&gt;&gt;
-Scanner
+ScannerV2
+--
+**Explanation:**
+This abstraction layer represents the user API (ROS-agnostic) and consists of the shown functions.
```
Comment (ct2034) [inline doc/diag_comp_overall_architecture.uxf:244]: "consists of the shown functions." is not necessary. Instead, write what it is doing.

File: doc/diag_comp_overall_architecture.uxf [hunk 9]
```diff
@@ -42,347 +67,405 @@ UDP package</panel_attributes>
...
+      <w>450</w>
+      <h>462</h>
+    </coordinates>
+    <panel_attributes>symbol=component
+style=wordwrap
+layer=-10
+Scanner protocol layer
+--
+**Explanation:**
+This abstraction layer implements the scanner protocol. In other words, this layer controls and organizes when and what data are sent and/or received from the scanner.
```
Comment (ct2034) [inline doc/diag_comp_overall_architecture.uxf:455]:
```suggestion
This abstraction layer implements the scanner protocol. In other words, this layer controls and organizes when and what data is sent and/or received from the scanner.
```

File: include/psen_scan_v2/raw_processing.h [hunk 10]
```diff
@@ -25,8 +25,15 @@
 
 namespace psen_scan_v2
 {
+/**
+ * @brief Namespace containing functions, etc. needed to transform
+ * higher level data to raw data (sent to or received from scanner) or vice versa.
```
Comment (ct2034) [inline include/psen_scan_v2/raw_processing.h:30]: Data data data data?
```suggestion
 * higher level interpretation to raw data (sent to or received from scanner) or vice versa.
```

File: doc/diag_comp_overall_architecture.uxf [hunk 11]
```diff
@@ -42,347 +67,405 @@ UDP package</panel_attributes>
...
-      <x>504</x>
-      <y>406</y>
-      <w>49</w>
-      <h>28</h>
-    </coordinates>
-    <panel_attributes>lt=..
+ScannerProtocolDef
+--
+**Explanation:**
+This component represents the scanner protocol/ state machine (indirectly) described in the Communication manual of the scanner on page 1.
```
Comment (ct2034) [inline doc/diag_comp_overall_architecture.uxf:222]:
```suggestion
This component represents the scanner protocol / state machine (indirectly) described in the communication manual of the scanner on page 1.
```

Comment (ct2034) [reply doc/diag_comp_overall_architecture.uxf:222]: And: where can I find this manual? Is there a link somewhere? Does the user even have access to this?

File: doc/diag_comp_overall_architecture.uxf [hunk 12]
```diff
@@ -42,347 +67,405 @@ UDP package</panel_attributes>
...
+      <x>384</x>
+      <y>384</y>
+      <w>378</w>
+      <h>348</h>
     </coordinates>
     <panel_attributes>symbol=component
+style=wordwrap
+layer=0
+group=1
 &lt;&lt;State machine&gt;&gt;
```
Comment (ct2034) [inline doc/diag_comp_overall_architecture.uxf:218]: what does this meand? why are there << >> ?

File: include/psen_scan_v2/diagnostics.h [hunk 13]
```diff
@@ -123,6 +130,17 @@ static const std::map<ErrorType, ErrorMessage> error_code_to_string
   }};
 // clang-format on
 
+/**
+ * @brief Defines a byte and bit position of an error in the diagnostic chunk.
```
Comment (ct2034) [inline include/psen_scan_v2/diagnostics.h:134]:
```suggestion
 * @brief Defines the byte and bit position of the error in the diagnostic chunk.
```

File: include/psen_scan_v2/diagnostics.h [hunk 14]
```diff
@@ -123,6 +130,17 @@ static const std::map<ErrorType, ErrorMessage> error_code_to_string
...
+ *
+ * The diagnostic chunk provided by the scanner when diagnostics are enabled in the ScannerConfiguration
+ * is a set of error bits stored in consecutive bytes.
+ *
+ * This class helps defining the positions of those error bits.
```
Comment (ct2034) [inline include/psen_scan_v2/diagnostics.h:139]:
```suggestion
 *
```

File: include/psen_scan_v2/diagnostics.h [hunk 15]
```diff
@@ -143,6 +161,17 @@ class ErrorLocation
   BitLocation bit_;
 };
 
+/**
+ * @brief Defines an Diagnostic message by defining the ErrorLocation and a scanner ID
```
Comment (ct2034) [inline include/psen_scan_v2/diagnostics.h:165]:
```suggestion
 * @brief Defines a diagnostic message, where it occurred in `ErrorLocation` and with which `ScannerId`
```

File: include/psen_scan_v2/laserscan.h [hunk 16]
```diff
@@ -23,21 +23,27 @@
 
 namespace psen_scan_v2
 {
-//! @brief Holds the measurement data for one laserscan.
+/**
+ *  @brief This class represents a single laser scan in the \<prefix\>_scan target frame.
```
Comment (ct2034) [inline include/psen_scan_v2/laserscan.h:27]:
```suggestion
 *  @brief Represents a single laser scan in the \<prefix\>_scan target frame.
```

File: include/psen_scan_v2/laserscan.h [hunk 17]
```diff
@@ -23,21 +23,27 @@
...
+ *
+ * All relevant information about a single scan of a psen_scan scanner are stored in this class.
+ * This includes the following information:
+ *
+ * - Distance measures in mm.
+ * - Normalized intensities of the signals.
+ * - Resolution of the scan in radians.
+ * - Min and Max angle in radians.
+ *
+ * The measures use the target frame defined as \<prefix\>_scan.
```
Comment (ct2034) [inline include/psen_scan_v2/laserscan.h:37]:
```suggestion
 * The measurements use the target frame defined as \<prefix\>_scan.
```

File: include/psen_scan_v2/laserscan_conversions.h [hunk 18]
```diff
@@ -22,6 +22,14 @@
 
 namespace psen_scan_v2
 {
+/**
+ * @brief Converts the technical monitoring_frame to the user friendly LaserScan type sent by the
+ * IScanner::LaserScanCallback.
+ * @see LaserScan
+ * @see monitoring_frame::Message
+ * @see ScannerV2
+ * @see IScanner
```
Comment (ct2034) [inline include/psen_scan_v2/laserscan_conversions.h:31]:
```suggestion
 * @see monitoring_frame::Message
 * @see ScannerV2
```
those are already linked anyways

File: include/psen_scan_v2/monitoring_frame_deserialization.h [hunk 19]
```diff
@@ -24,6 +24,20 @@
 
 namespace psen_scan_v2
 {
+/**
+ * @brief Namespace defining all information/types needed to describe a monitoring frame.
```
Comment (ct2034) [inline include/psen_scan_v2/monitoring_frame_deserialization.h:28]:
```suggestion
 * @brief Defining all information and types needed to describe a monitoring frame.
```

File: include/psen_scan_v2/monitoring_frame_deserialization.h [hunk 20]
```diff
@@ -24,6 +24,20 @@
...
+ *
+ * A complete scan round can be divided in several parts sent independently over the UDP connection.
```
Comment (ct2034) [inline include/psen_scan_v2/monitoring_frame_deserialization.h:30]:
```suggestion
 * The scanner splits up a full rotation in several parts and sends them in separate parts and sends them via UDP packages.
```

File: include/psen_scan_v2/monitoring_frame_deserialization.h [hunk 21]
```diff
@@ -24,6 +24,20 @@
...
+ * Those parts are called monitoring frames.
+ *
+ * Every single frame **has to** contain some general information about the scan in the fixed fields
+ * and **can** contain additional data like distances or intensities in additional fields.
+ *
+ * @see FixedFields
+ * @see Header
+ * @see HeaderID
```
Comment (ct2034) [inline include/psen_scan_v2/monitoring_frame_deserialization.h:38]: these are no links ?

File: include/psen_scan_v2/monitoring_frame_deserialization.h [hunk 22]
```diff
@@ -71,8 +88,25 @@ class FixedFields
   FromTheta from_theta_;
   Resolution resolution_;
 };
+/**
+ * @brief Namespace containing all information/types needed to describe the content of the additional information field
```
Comment (ct2034) [inline include/psen_scan_v2/monitoring_frame_deserialization.h:92]:
```suggestion
 * @brief Contains all information and types for additional information field
```

**Review by ct2034 — CHANGES_REQUESTED** [review]

File: include/psen_scan_v2/ros_scanner_node.h [hunk 23]
```diff
@@ -34,7 +34,7 @@ namespace psen_scan_v2
 using namespace std::chrono_literals;
 
 /**
- * @brief ROS Node for fetching and publishing laserscan data from the scanner.
+ * @brief ROS Node that continuesly publishes scan data of a single PSENscan laser scanner.
```
Comment (jschleicher) [inline include/psen_scan_v2/ros_scanner_node.h:37]:
```suggestion
 * @brief ROS Node that continuously publishes scan data of a single PSENscan laser scanner.
```

By the way (not a documentation but design question): What does the `T` suffix in `RosScannerNodeT` mean?

File: include/psen_scan_v2/monitoring_frame_deserialization.h (hunk 22 above)
Comment (jschleicher) [reply include/psen_scan_v2/monitoring_frame_deserialization.h:92]: All namespace descriptions start with `Namespace containing` - so you could either keep it here as well or search-replace all of them ;-) ?

File: include/psen_scan_v2/ros_scanner_node.h (hunk 23 above)
Comment (rfeistenauer) [reply include/psen_scan_v2/ros_scanner_node.h:37]: Without being present at the decision I expect the T stands for Template.
```
typedef ROSScannerNodeT<> ROSScannerNode;

template <typename S>
ROSScannerNodeT<S>
```

I think the template method was mainly used to inject the scanner mock for the integration tests.

File: doc/calls_and_callbacks.md (hunk 7 above)
Comment (rfeistenauer) [reply doc/calls_and_callbacks.md:1]: You are referring to a line that was taken out, is this outdated or did you mean something else?

**Pull request merged**